students/julienjass/Lesson10/mailroom_fp.py

```diff
@@ -0,0 +1,297 @@
+
+    @property
+    def filename(self):
+        return self._name.replace(' ', '_') + '.txt'
+
+    def __repr__(self):
+        return "{}: {}".format(self._name, self._donations)  
+
+
+class DonorDB:
+    
+    def __init__(self):
+        self._donors = {}
+
+    def add_donor(self, donor):
+        if donor.name.lower() in self._donors:
+            self._donors[donor.name.lower()].add_donation(*(donor.donations))
+        else:
+            self._donors[donor.name.lower()] = donor
+
+    def get_total_from_donor(self, donor_name):
+        return self._donors[donor_name.lower()].total_donations
+
+    def get_donor(self, donor_name):
+        try:
+            return self._donors[donor_name.lower()]
+        except KeyError:
+            print('This Donor doesn''t exist in the database')
+
+    @property
+    def export(self):
+        return {donor.name:donor.donations for donor in self._donors.values()}
+
+    def donor_list(self):
+        name_list = [donor.name for donor in self._donors.values()]
+        return '\n'.join(name_list)
+
+    def letters_on_disk(self):
+        for donor in self._donors.values():
+            print('Creating a letter for {:s}'.format(donor.name))
+            with open(donor.filename, 'w') as outfile:
+                outfile.write(donor.letter)
+        print()
+
+    def report(self):
+        sorted_donors = [donor for donor in self._donors.values()]
+        sorted_donors.sort(key=lambda donor: donor.total_donations, reverse=True)
+        report_rows = []
+        for d in sorted_donors:
+            report_rows.append('{:26s} {:>12s} {:^13d} {:>12s}'.format(d.name, '${:,.2f}'.format(d.total_donations), d.count_donations, '${:,.2f}'.format(d.average_donation)))
+        header = ('Donor Name                | Total Given | Num Gifts | Average Gift\n') + ('-' * 66) + '\n'
+        return header + '\n'.join(report_rows) + '\n'
+
+    @property
+    def count_donors(self):
+        return len(self._donors)
+
+    @property
+    def count_donations(self):
+        return sum([len(donor.donations) for donor in self._donors.values()])
+
+    @property
+    def total_donations(self):
+        return sum([donor.total_donations for donor in self._donors.values()])
+
+    @property
+    def average_total_donation(self):
+        return self.total_donations / self.count_donors
+
+    @property
+    def average_single_donation(self):
+        return self.total_donations / self.count_donations
+
+    def __repr__(self):
+        return str([d for d in self._donors.values()])
+
+    @property
+    def projection(self):
+        return {d[0]:d[1] for d in self._projection}
+
+    def challenge(self, x=1, mn=0, mx=float("inf")):
+        self._projection = []
+        for k, v in self.export.items():
+            self._projection.append((k, list(map(lambda y: y * x, list(filter(lambda z: z >= mn and z <= mx, v))))))
```

Comment:
Although it's common to use 'throw away' variable names for things like lambdas, this code is pretty obscure. I would suggest using meaningful variable names to make it more readable